From 10389b5c9fe7f756509b757902ae6ef273a6a4db Mon Sep 17 00:00:00 2001
From: David S. Wang <dsw@cloudera.com>
Date: Tue, 28 Feb 2012 14:23:43 -0800
Subject: [PATCH 054/154] HBASE-5484 Fix spelling mistakes in error message when starting HBase in standalone mode and the configured ZK port is not able to be bound to.

---
 .../hadoop/hbase/master/HMasterCommandLine.java    |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java b/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
index 44d28df..66ee021 100644
--- a/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
+++ b/src/main/java/org/apache/hadoop/hbase/master/HMasterCommandLine.java
@@ -125,9 +125,10 @@ public class HMasterCommandLine extends ServerCommandLine {
         zooKeeperCluster.setDefaultClientPort(zkClientPort);
         int clientPort = zooKeeperCluster.startup(zkDataPath);
         if (clientPort != zkClientPort) {
-          String errorMsg = "Couldnt start ZK at requested address of " +
-            zkClientPort + ", instead got: " + clientPort + ". Aborting. Why? " +
-            "Because clients (eg shell) wont be able to find this ZK quorum";
+          String errorMsg = "Could not start ZK at requested port of " +
+            zkClientPort + ".  ZK was started at port: " + clientPort +
+            ".  Aborting as clients (e.g. shell) will not be able to find " +
+            "this ZK quorum.";
           System.err.println(errorMsg);
           throw new IOException(errorMsg);
         }
-- 
1.7.0.4

