From 1e39628cf189db531e319c31ceb23c3c3cb0505f Mon Sep 17 00:00:00 2001
From: Shreepadma Venugopalan <shreepadma@cloudera.com>
Date: Thu, 10 Jan 2013 15:07:26 -0800
Subject: [PATCH 022/121] CDH-9602. restrict number of bit vectors from Hive CLI, to prevent out of Java heap memory

---
 .../ql/udf/generic/GenericUDAFComputeStats.java    |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/src/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFComputeStats.java b/src/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFComputeStats.java
index e09efb7..857d538 100644
--- a/src/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFComputeStats.java
+++ b/src/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFComputeStats.java
@@ -475,6 +475,10 @@ public class GenericUDAFComputeStats extends AbstractGenericUDAFResolver {
 
       if (myagg.firstItem) {
         int numVectors = PrimitiveObjectInspectorUtils.getInt(parameters[1], numVectorsOI);
+        if (numVectors > 1024) {
+          throw new HiveException("The maximum allowed value for number of bit vectors " +
+            " is 1024, but was passed " + numVectors + " bit vectors");
+        }
         initNDVEstimator(myagg, numVectors);
         myagg.firstItem = false;
         myagg.numBitVectors = numVectors;
@@ -636,6 +640,7 @@ public class GenericUDAFComputeStats extends AbstractGenericUDAFResolver {
       if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {
         inputOI = (PrimitiveObjectInspector) parameters[0];
         numVectorsOI = (PrimitiveObjectInspector) parameters[1];
+           
       } else {
         soi = (StructObjectInspector) parameters[0];
 
@@ -773,6 +778,10 @@ public class GenericUDAFComputeStats extends AbstractGenericUDAFResolver {
 
       if (myagg.firstItem) {
         int numVectors = PrimitiveObjectInspectorUtils.getInt(parameters[1], numVectorsOI);
+        if (numVectors > 1024) {
+          throw new HiveException("The maximum allowed value for number of bit vectors " +
+            " is 1024, but was passed " + numVectors + " bit vectors");
+        }
         initNDVEstimator(myagg, numVectors);
         myagg.firstItem = false;
         myagg.numBitVectors = numVectors;
@@ -1085,6 +1094,10 @@ public class GenericUDAFComputeStats extends AbstractGenericUDAFResolver {
 
       if (myagg.firstItem) {
         int numVectors = PrimitiveObjectInspectorUtils.getInt(parameters[1], numVectorsOI);
+        if (numVectors > 1024) {
+          throw new HiveException("The maximum allowed value for number of bit vectors " +
+            " is 1024, but was passed " + numVectors + " bit vectors");
+        }
         initNDVEstimator(myagg, numVectors);
         myagg.firstItem = false;
         myagg.numBitVectors = numVectors;
-- 
1.7.0.4

