From 8b06fe69cdf55c5a35c60c7e4bf376de3d231905 Mon Sep 17 00:00:00 2001
From: Shreepadma Venugopalan <shreepadma@cloudera.com>
Date: Thu, 10 Jan 2013 18:11:03 -0800
Subject: [PATCH 024/121] CDH-8941. hive.server2 property is incorrect in hive-default.xml.template file

---
 conf/hive-default.xml.template |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/conf/hive-default.xml.template b/src/conf/hive-default.xml.template
index 89b6a60..aaf09b9 100644
--- a/src/conf/hive-default.xml.template
+++ b/src/conf/hive-default.xml.template
@@ -1646,7 +1646,7 @@
 </property>
 
 <property>
-  <name>hive.server2.kerberos.principal</name>
+  <name>hive.server2.authentication.kerberos.principal</name>
   <value></value>
   <description>
     Kerberos server principal
@@ -1654,7 +1654,7 @@
 </property>
 
 <property>
-  <name>hive.server2.kerberos.keytab</name>
+  <name>hive.server2.authentication.kerberos.keytab</name>
   <value></value>
   <description>
     Kerberos keytab file for server principal
-- 
1.7.0.4

