From f3befbb799147a13dc186031a40c63116ea0dd22 Mon Sep 17 00:00:00 2001
From: Prasad Mujumdar <prasadm@cloudera.com>
Date: Tue, 5 Feb 2013 13:05:15 -0800
Subject: [PATCH 056/121] CDH-8578:  Secured HiveServer2 logs-in after connecting to HiveMetaStore. Fails if HMS is also kerberized. Additional patch to support _HOST alias

---
 .../hadoop/hive/shims/HadoopShimsSecure.java       |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/src/shims/src/common-secure/java/org/apache/hadoop/hive/shims/HadoopShimsSecure.java b/src/shims/src/common-secure/java/org/apache/hadoop/hive/shims/HadoopShimsSecure.java
index 9968825..cae0d82 100644
--- a/src/shims/src/common-secure/java/org/apache/hadoop/hive/shims/HadoopShimsSecure.java
+++ b/src/shims/src/common-secure/java/org/apache/hadoop/hive/shims/HadoopShimsSecure.java
@@ -59,6 +59,7 @@ import org.apache.hadoop.mapred.TaskID;
 import org.apache.hadoop.mapred.lib.CombineFileInputFormat;
 import org.apache.hadoop.mapred.lib.CombineFileSplit;
 import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.security.SecurityUtil;
 import org.apache.hadoop.security.UserGroupInformation;
 import org.apache.hadoop.security.token.Token;
 import org.apache.hadoop.security.token.TokenIdentifier;
@@ -558,7 +559,8 @@ public abstract class HadoopShimsSecure implements HadoopShims {
 
   @Override
   public void loginUserFromKeytab(String principal, String keytabFile) throws IOException {
-    UserGroupInformation.loginUserFromKeytab(principal, keytabFile);
+    String hostPrincipal = SecurityUtil.getServerPrincipal(principal, "0.0.0.0");
+    UserGroupInformation.loginUserFromKeytab(hostPrincipal, keytabFile);
   }
 
   @Override
-- 
1.7.0.4

