From 9df65033b2c8499984b461430fa68cfefd70c29c Mon Sep 17 00:00:00 2001
From: Prasad Mujumdar <prasadm@cloudera.com>
Date: Fri, 15 Mar 2013 12:51:51 -0700
Subject: [PATCH 074/121] CDH-11031: TExecuteStatement() creates a NullPointerException

---
 .../cli/operation/ExecuteStatementOperation.java   |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/src/service/src/java/org/apache/hive/service/cli/operation/ExecuteStatementOperation.java b/src/service/src/java/org/apache/hive/service/cli/operation/ExecuteStatementOperation.java
index a25dbda..c415f79 100644
--- a/src/service/src/java/org/apache/hive/service/cli/operation/ExecuteStatementOperation.java
+++ b/src/service/src/java/org/apache/hive/service/cli/operation/ExecuteStatementOperation.java
@@ -46,10 +46,6 @@ public abstract class ExecuteStatementOperation extends Operation {
     String[] tokens = statement.trim().split("\\s+");
     String command = tokens[0].toLowerCase();
 
-    for (Entry<String, String> opConf : confOverlay.entrySet()) {
-      parentSession.getHiveConf().set(opConf.getKey(), opConf.getValue());
-    }
-
     ExecuteStatementOperation newExecOP;
     if ("set".equals(command)) {
       newExecOP = new SetOperation(parentSession, statement, confOverlay);
-- 
1.7.0.4

