From 4f81aaa514f05d381ab9117cddbbb4eee426b1a0 Mon Sep 17 00:00:00 2001
From: Esteban Gutierrez <esteban@cloudera.com>
Date: Tue, 3 Feb 2015 09:16:38 -0800
Subject: [PATCH 043/110] CLOUDERA-BUILD remove TestAccessController @Test annotation in verifyAnyCreate

---
 .../security/access/TestAccessController.java      |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestAccessController.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestAccessController.java
index 5aa8a4b..7949035 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestAccessController.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/security/access/TestAccessController.java
@@ -2480,7 +2480,6 @@ public class TestAccessController extends SecureTestUtil {
     TEST_UTIL.getMiniHBaseCluster().getMaster().deleteNamespace(ns);
   }
 
-  @Test
   private void verifyAnyCreate(AccessTestAction action) throws Exception {
     verifyAllowed(action, SUPERUSER, USER_ADMIN, USER_OWNER, USER_CREATE, USER_ADMIN_CF);
     verifyDenied(action, USER_NONE, USER_RO, USER_RW);
-- 
1.7.0.4

