? openssl.m4.libsocket.diff
Index: openssl.m4
===================================================================
RCS file: /afs/andrew/system/cvs/src/cmulocal/openssl.m4,v
retrieving revision 1.11
diff -u -r1.11 openssl.m4
--- openssl.m4	17 May 2006 18:30:19 -0000	1.11
+++ openssl.m4	27 Apr 2009 17:36:01 -0000
@@ -33,7 +33,8 @@
 		AC_CHECK_HEADER(openssl/evp.h, [
 			AC_CHECK_LIB(crypto, EVP_DigestInit,
 					with_openssl="yes",
-					with_openssl="no", $LIB_RSAREF)],
+					with_openssl="no",
+					$LIB_RSAREF $LIB_SOCKET)],
 			with_openssl=no)
 		;;
 esac
