From 93f769181d59e8cf98c37762261bc1474c181ab8 Mon Sep 17 00:00:00 2001
From: Todd Lipcon <todd@cloudera.com>
Date: Wed, 12 Jan 2011 13:32:54 -0800
Subject: [PATCH 03/18] HBASE-3533. Allow HBASE_LIBRARY_PATH to point to libs

Author: Todd Lipcon
---
 bin/hbase |   19 ++++++++++++-------
 1 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/bin/hbase b/bin/hbase
index 347fba6..f3b872b 100755
--- a/bin/hbase
+++ b/bin/hbase
@@ -181,21 +181,26 @@ if $cygwin; then
   HBASE_HOME=`cygpath -d "$HBASE_HOME"`
   HBASE_LOG_DIR=`cygpath -d "$HBASE_LOG_DIR"`
 fi
+
+function append_path() {
+  if [ -z "$1" ]; then
+    echo $2
+  else
+    echo $1:$2
+  fi
+}
+
 # setup 'java.library.path' for native-hadoop code if necessary
-JAVA_LIBRARY_PATH=''
+JAVA_LIBRARY_PATH="$HBASE_LIBRARY_PATH"
 if [ -d "${HBASE_HOME}/build/native" -o -d "${HBASE_HOME}/lib/native" ]; then
   JAVA_PLATFORM=`CLASSPATH=${CLASSPATH} ${JAVA} org.apache.hadoop.util.PlatformName | sed -e "s/ /_/g"`
 
   if [ -d "$HBASE_HOME/build/native" ]; then
-    JAVA_LIBRARY_PATH=${HBASE_HOME}/build/native/${JAVA_PLATFORM}/lib
+    JAVA_LIBRARY_PATH=$(append_path "$JAVA_LIBRARY_PATH" ${HBASE_HOME}/build/native/${JAVA_PLATFORM}/lib)
   fi
 
   if [ -d "${HBASE_HOME}/lib/native" ]; then
-    if [ "x$JAVA_LIBRARY_PATH" != "x" ]; then
-      JAVA_LIBRARY_PATH=${JAVA_LIBRARY_PATH}:${HBASE_HOME}/lib/native/${JAVA_PLATFORM}
-    else
-      JAVA_LIBRARY_PATH=${HBASE_HOME}/lib/native/${JAVA_PLATFORM}
-    fi
+    JAVA_LIBRARY_PATH=$(append_path "$JAVA_LIBRARY_PATH" ${HBASE_HOME}/lib/native/${JAVA_PLATFORM})
   fi
 fi
 
-- 
1.7.0.4

