From cb38cfcc2fc13dcbc314ed9b3e742e6db6c9121a Mon Sep 17 00:00:00 2001
From: Jarek Jarcec Cecho <jarcec@cloudera.com>
Date: Thu, 14 Feb 2013 14:08:54 -0800
Subject: [PATCH 66/67] HIVE-4021 PostgreSQL upgrade scripts are creating column with incorrect name

---
 .../upgrade/postgres/012-HIVE-1362.postgres.sql    |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/metastore/scripts/upgrade/postgres/012-HIVE-1362.postgres.sql b/src/metastore/scripts/upgrade/postgres/012-HIVE-1362.postgres.sql
index 1914b5f..89314ed 100644
--- a/src/metastore/scripts/upgrade/postgres/012-HIVE-1362.postgres.sql
+++ b/src/metastore/scripts/upgrade/postgres/012-HIVE-1362.postgres.sql
@@ -10,7 +10,7 @@ CREATE TABLE "TAB_COL_STATS" (
  "LONG_LOW_VALUE" bigint,
  "LONG_HIGH_VALUE" bigint,
  "DOUBLE_LOW_VALUE" double precision,
- "DOUBLE_HIGH_VALUES" double precision,
+ "DOUBLE_HIGH_VALUE" double precision,
  "BIG_DECIMAL_LOW_VALUE" character varying(4000) DEFAULT NULL::character varying,
  "BIG_DECIMAL_HIGH_VALUE" character varying(4000) DEFAULT NULL::character varying,
  "NUM_NULLS" bigint NOT NULL,
@@ -38,7 +38,7 @@ CREATE TABLE "PART_COL_STATS" (
  "LONG_LOW_VALUE" bigint,
  "LONG_HIGH_VALUE" bigint,
  "DOUBLE_LOW_VALUE" double precision,
- "DOUBLE_HIGH_VALUES" double precision,
+ "DOUBLE_HIGH_VALUE" double precision,
  "BIG_DECIMAL_LOW_VALUE" character varying(4000) DEFAULT NULL::character varying,
  "BIG_DECIMAL_HIGH_VALUE" character varying(4000) DEFAULT NULL::character varying,
  "NUM_NULLS" bigint NOT NULL,
-- 
1.7.0.4

