From b838d9d22f1eb6a7b2160043d1b62b5952c197ac Mon Sep 17 00:00:00 2001
From: Carl Steinbach <carl@cloudera.com>
Date: Fri, 2 Mar 2012 12:39:00 -0800
Subject: [PATCH 31/58] CLOUDERA-BUILD. Update Eclipse classpath template

---
 eclipse-templates/.classpath |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/src/eclipse-templates/.classpath b/src/eclipse-templates/.classpath
index f3a7d14..d3c5d03 100644
--- a/src/eclipse-templates/.classpath
+++ b/src/eclipse-templates/.classpath
@@ -181,6 +181,7 @@
   <classpathentry kind="src" path="jdbc/src/test"/>
   <classpathentry kind="src" path="shims/src/@HADOOPVERPREF@/java"/>
   <classpathentry kind="src" path="shims/src/common/java"/>
+  <classpathentry kind="src" path="shims/src/common-secure/java"/>
   <classpathentry kind="src" path="hwi/src/java"/>
   <classpathentry kind="src" path="hwi/src/test"/>
   <classpathentry kind="src" path="hbase-handler/src/java"/>
-- 
1.7.0.4

